site/learn/BestPractice-Versioning.md

```diff
@@ -0,0 +1,115 @@
+
+Let’s understand by an example. 
+We have a schema named user, which had fields such as id, firstName and lastName. 
+```graphql
+type User {
+	id: Int!,
+	firstName : String!,
+	lastName : String!
+}
+```
+
+In the above schema all fields are required. But in the next version we can add a variable and can also change the requirements. Such as :
+```graphql
+type User {
+	Id : Int!,
+	firstName : String!,
+	middleName : String!
+	lastName : String,
+}
+```
+
+Here we have added the middleName parameter which is mandatory and have changed lastName to optional.
+
+
+* The best practice is to use “@deprecated(reason)” defined in the schema. By using this the client can know which filed to use and which one is deprecated. This will not define the version number as the semantic versioning does. The deprecated fields will be shown by the GraphQL Explorer.
+
+
+### More info on Versioning
+There are few other best practices for versioning GraphQL APIs.
+Version constraints can be added to the query which will tell that the field will be resolved by which version. Using version constraints we can access different fields of different versions. For example, the query will look like : 
+```graphql
+query {
+	account(id : 5){
+		old : profilePhoto(versionContraint : “^0.1”)
+		new : profilePhoto(versionContraint : “^0.2”)
+	}
+}
+```
+
+We can define the versions in the schema and query can be written as above to access the fields of different versions.
+
+### Evolution of Schema
+This is definitely different from versioning, but this will be helpful to build schemas which can be evolved with time without making breaking changes. Schema can be evolved by making changes such as adding/removing fields, structures, types, etc. While making these changes, there can be some points which will restrict the schema while evolving and then @deprecated needs to be used as a last resort. For example there is a schema :
+```graphql
+type  User{
+	id: String!
+	name: String!
+	Username : String!
+	birthDate : DateTime!
+	lastLoginHistory : [DateTime]!
+}
+
+```
+In the above schema we have one field named lastLoginHistory which stores the DateTime fields as a list. Now there needs some changes in the schema and some additional data needs to be added in lastLoginHistory. For that we need to add another field to the User schema. These changes will add many fields and this could make the schema lengthier and worse.
+
+But rather than this, we can build a schema for storing lastLoginHistory data and we can use that schema referenced to the User schema.
+```graphql
+type NewDateTime : {
+	lastLoginTime : DateTime!
+	lastLoginPeriodLength : Time!
+}
+```
+``` graphql
+type User{
+	id: String!
+	name: String!
+	Username : String!
+	birthDate : DateTime!
+	lastLoginHistory : NewDateTime!
```

Comment:
but... this would be a breaking change for clients?

Reply:
How? I think we can evolve schema by referencing schemas to others. Any specifics why this will be a breaking change?